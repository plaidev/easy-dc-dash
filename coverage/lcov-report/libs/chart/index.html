<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for libs/chart</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> libs/chart
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">41.94% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>164/391</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">33.03% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>109/330</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">45.05% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>41/91</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">45.09% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>147/326</span>
      </div>
    </div>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="_base.js"><a href="_base.js.html">_base.js</a></td>
	<td data-value="37.2" class="pic low"><div class="chart"><div class="cover-fill" style="width: 37%;"></div><div class="cover-empty" style="width:63%;"></div></div></td>
	<td data-value="37.2" class="pct low">37.2%</td>
	<td data-value="336" class="abs low">125/336</td>
	<td data-value="28.52" class="pct low">28.52%</td>
	<td data-value="277" class="abs low">79/277</td>
	<td data-value="39.24" class="pct low">39.24%</td>
	<td data-value="79" class="abs low">31/79</td>
	<td data-value="39.43" class="pct low">39.43%</td>
	<td data-value="279" class="abs low">110/279</td>
	</tr>

<tr>
	<td class="file medium" data-value="_coordinateGridBase.js"><a href="_coordinateGridBase.js.html">_coordinateGridBase.js</a></td>
	<td data-value="65.96" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 65%;"></div><div class="cover-empty" style="width:35%;"></div></div></td>
	<td data-value="65.96" class="pct medium">65.96%</td>
	<td data-value="47" class="abs medium">31/47</td>
	<td data-value="56.6" class="pct medium">56.6%</td>
	<td data-value="53" class="abs medium">30/53</td>
	<td data-value="81.82" class="pct high">81.82%</td>
	<td data-value="11" class="abs high">9/11</td>
	<td data-value="74.36" class="pct medium">74.36%</td>
	<td data-value="39" class="abs medium">29/39</td>
	</tr>

<tr>
	<td class="file high" data-value="ordinal-bar.vue"><a href="ordinal-bar.vue.html">ordinal-bar.vue</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Nov 30 2017 17:42:44 GMT+0900 (JST)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
