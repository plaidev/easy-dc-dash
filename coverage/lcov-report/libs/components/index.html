<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for libs/components</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> libs/components
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">56.25% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>36/64</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">50% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>20/40</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">56.52% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>13/23</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">56.25% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>36/64</span>
      </div>
    </div>
  </div>
  <div class='status-line medium'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="card.vue"><a href="card.vue.html">card.vue</a></td>
	<td data-value="73.91" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 73%;"></div><div class="cover-empty" style="width:27%;"></div></div></td>
	<td data-value="73.91" class="pct medium">73.91%</td>
	<td data-value="46" class="abs medium">34/46</td>
	<td data-value="58.82" class="pct medium">58.82%</td>
	<td data-value="34" class="abs medium">20/34</td>
	<td data-value="73.33" class="pct medium">73.33%</td>
	<td data-value="15" class="abs medium">11/15</td>
	<td data-value="73.91" class="pct medium">73.91%</td>
	<td data-value="46" class="abs medium">34/46</td>
	</tr>

<tr>
	<td class="file low" data-value="chart-link.vue"><a href="chart-link.vue.html">chart-link.vue</a></td>
	<td data-value="12.5" class="pic low"><div class="chart"><div class="cover-fill" style="width: 12%;"></div><div class="cover-empty" style="width:88%;"></div></div></td>
	<td data-value="12.5" class="pct low">12.5%</td>
	<td data-value="8" class="abs low">1/8</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="3" class="abs low">1/3</td>
	<td data-value="12.5" class="pct low">12.5%</td>
	<td data-value="8" class="abs low">1/8</td>
	</tr>

<tr>
	<td class="file low" data-value="krt-dc-tooltip.vue"><a href="krt-dc-tooltip.vue.html">krt-dc-tooltip.vue</a></td>
	<td data-value="11.11" class="pic low"><div class="chart"><div class="cover-fill" style="width: 11%;"></div><div class="cover-empty" style="width:89%;"></div></div></td>
	<td data-value="11.11" class="pct low">11.11%</td>
	<td data-value="9" class="abs low">1/9</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="4" class="abs low">1/4</td>
	<td data-value="11.11" class="pct low">11.11%</td>
	<td data-value="9" class="abs low">1/9</td>
	</tr>

<tr>
	<td class="file low" data-value="reset-button.vue"><a href="reset-button.vue.html">reset-button.vue</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Nov 30 2017 17:42:44 GMT+0900 (JST)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
