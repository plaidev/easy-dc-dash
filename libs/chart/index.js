import Base from './_base'
import {compose} from './_composite'
import SegmentPie from './segment-pie.vue'
import WeekRow from './week-row.vue'
import ListRow from './list-row.vue'
import RateLine from './rate-line.vue'
import StackedLines from './stacked-lines.vue'
import OrdinalBar from './ordinal-bar.vue'
import StackedBar from './stacked-bar.vue'
import FilterStackedBar from './filter-stacked-bar.vue'
import GeoJP from './geo-jp.vue'
import DataTable from './data-table.vue'
import resetAllButton from './reset-all-button.vue'

const components = {
  'segment-pie': SegmentPie,
  'week-row': WeekRow,
  'list-row': ListRow,
  'rate-line': RateLine,
  'stacked-lines': StackedLines,
  'ordinal-bar': OrdinalBar,
  'stacked-bar': StackedBar,
  'filter-stacked-bar': FilterStackedBar,
  'geo-jp': GeoJP,
  'data-table': DataTable,
  'stack-and-rate': compose(StackedLines, RateLine),
  'reset-all-button': resetAllButton
}

function install(Vue, options) {
  Vue.mixin({
    components: components
  })
}

export default {
  Base: Base,
  RateLine: RateLine,
  StackedLines: StackedLines,
  WeekRow: WeekRow,
  ListRow: ListRow,
  SegmentPie: SegmentPie,
  OrdinalBar: OrdinalBar,
  StackedBar: StackedBar,
  FilterStackedBar: FilterStackedBar,
  GeoJP: GeoJP,
  DataTable: DataTable,
  compose: compose,
  resetAllButton: resetAllButton,
  install: install,
  installedComponents: components
}
