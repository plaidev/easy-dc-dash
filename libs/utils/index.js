import dc from 'dc'

export function generateUUID() {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
    var r = Math.random()*16|0, v = c == 'x' ? r : (r&0x3|0x8);
    return v.toString(16);
  })
}


export function generateDomId() {
  return 'id-' + generateUUID();
}


// String(js): <data-table columns="{key: d.key}"></data-table>
// Object: <data-table :columns="{key: "key"}"></data-table>
// Array: <data-table :columns="["key"]"></data-table>
// Function: <data-table :columns="customParser"></data-table>
export function generateExtractor (rule) {
  if (typeof rule === 'function' || rule instanceof Function) {
    return rule
  }

  else if (typeof rule === "string" || rule instanceof String) {
    return new Function('d', `const v = ${rule}; return v === null? "": v;`)
  }

  else if (rule instanceof Array) {
    return (d) => {
      row = {}
      rule.forEach((k) => {
        row[k] = d[rule[k]]
      })
      return row === null? '': row;
    }
  }

  else if (rule instanceof Object) {
    return (d) => {
      row = {}
      Object.keys(rule).forEach((k) => {
        row[k] = d[rule[k]]
      })
      return row === null? '': row;
    }
  }

  return // else
}

// https://github.com/dc-js/dc.js/wiki/FAQ#combine-groups
export function combineGroups(sourceGroups) {
  return {
    all: () => {
      const alls = sourceGroups.map((g) => g.all());
      // Object型がkeyになっている場合に ret.push({key: k, value:gm[k]}) のkeyがString型になってしまうのを防ぐ
      // ret.push({key: objKeys[k], value: gm[k]});
      const objKeys = {};
      const gm = {};
      alls.forEach((a, i) => {
        a.forEach((b) => {
          if(!gm[b.key]) {
            gm[b.key] = new Array(sourceGroups.length);
            for(let j=0; j<sourceGroups.length; ++j)
              gm[b.key][j] = 0;
          }
          gm[b.key][i] = b.value;
          for(let k in gm)
            if(!objKeys[k])
              objKeys[k] = b.key;
        });
      });
      const ret = [];
      for(let k in gm)
        ret.push({key: objKeys[k], value: gm[k]});
      return ret;
    }
  };
}

// https://github.com/dc-js/dc.js/wiki/FAQ#remove-empty-bins
export function removeEmptyBins(sourceGroup) {
  return {
    all: () => {
      return sourceGroup.all().filter((d) => d.value != 0)
    }
  };
}

// reverse dc.legend() order
// See: http://stackoverflow.com/questions/39811210/dc-charts-change-legend-order
export function reverseLegendOrder(chart) {
  dc.override(chart, 'legendables', function() {
      var items = chart._legendables();
      return items.reverse();
  });
}