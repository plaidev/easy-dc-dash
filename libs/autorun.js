import Vue from 'vue/dist/vue.js'
import {loadMode, loadCSV} from './utils/load-dataset'
import Store from './store'
import Chart from './chart'


function load() {

  const elms = document.querySelectorAll('easy-dc-dataset');

  if (elms.length <= 0) return Promise.reject();

  const promises = Array.prototype.map.call(elms, (el) => {
    const name = el.getAttribute('dataset') || undefined;
    const mode = el.getAttribute('mode');
    const csv = el.getAttribute('csv');

    const dateFields = el.getAttribute('date-fields')
    const dateFormat = el.getAttribute('date-format')
    const isUTC = el.getAttribute('date-is-utc')

    const options = {
      dateFields: dateFields ? dateFields.split(','): undefined,
      dateFormat,
      isUTC
    }

    let p;

    if (mode) {
      p = loadMode(mode, options)
        .then((data) => Store.registerData(data, {name}))
    }
    else if (csv) {
      p = loadCSV(csv, options)
        .then((data) => Store.registerData(data, {name}))
    }
    else {
      p = Promise.resolve()
    }

    return p
  });

  return Promise.all(promises)
}

function start() {
  const tags = Object.keys(Chart.installedComponents);
  tags.forEach((tag) => {
    document.querySelectorAll(tag).forEach((el) => {
      new Vue({el, data: Store.state.binds})
    });
  })
}

function _run() {
  Chart.install(Vue)
  load()
    .then(start)
    .catch(() => console.log('dataset setting not found. disable autorun.'))
}

export function run() {

  if (['complete', 'interactive'].includes(document.readyState)) {
    _run();
  }
  else {
    document.addEventListener("DOMContentLoaded",function(){
      _run();
    }, false);
  }
}

