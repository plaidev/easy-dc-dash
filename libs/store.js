import d3 from 'd3'
import crossfilter from 'crossfilter'
import dc from 'dc'
import {downloadCSV} from './utils/blob-csv'

//-------------------------------------


class DashboardStore {
  constructor() {
    this.state = {
      binds: {}
    };

    this._cf = {};
    this._charts = {};
    this._volumeBind = {};
    this._dimensions = {
      default: {}
    };
    this._labels = {};
  }

  setBindData(name, data) {
    this.state.binds[name] = data;
  }

  registerData(data=[], options={}) {
    const {
      dataset = 'default',
      labels = {}
    } = options;

    // crossfilterのインスタンス作成
    this._cf[dataset] = crossfilter(data);

    this.setLabels(labels, {dataset})
  }

  registerDimension(name, method, options={}) {
    const {
      dataset = 'default'
    } = options;

    if (!(dataset in this._dimensions)) {
      this._dimensions[dataset] = {};
    }

    // TODO: dimension作成数のlimit管理
    if (!(name in this._dimensions[dataset])) {
      this._dimensions[dataset][name] = this._cf[dataset].dimension(method)
    }
    return this._dimensions[dataset][name];
  }

  unregisterDimension(name, {dataset='default'}) {
    // TODO: implement
  }

  getDimension(name, options={}) {
    const {
      dataset = 'default'
    } = options;
    return this._dimensions[dataset][name];
  }

  getCfSize(options={}) {
    const {
      name = 'default'
    } = options;
    return this._cf[name].size();
  }

  registerChart(parent, name, chartType, binds={}) {
    const chart = new dc[chartType](parent);

    // volumeとして参照するchartがあれば登録
    if (binds.volume) {
      if (!this._volumeBind[binds.volume])
        this._volumeBind[binds.volume] = []
      this._volumeBind[binds.volume].push(name)

      if (this._charts[binds.volume]) {
        chart.rangeChart(this._charts[binds.volume])
      }
    }

    // このchartをvolumeとして参照するchartがあれば、bind
    if (this._volumeBind[name]) {
      this._volumeBind[name].forEach((refChart) => {
        this._charts[refChart].rangeChart(chart)
      })
    }

    this._charts[name] = chart;

    return chart;
  }

  unregisterChart(name, chart) {
    // TODO: implement
    // this._charts[name] = chart;
  }

  setLabels(labels, options={}) {
    const {
      dataset = 'default'
    } = options;
    if (!this._labels[dataset]) this._labels[dataset] = {};
    Object.assign(this._labels[dataset], labels);
  }

  getLabel(k, options={}) {
    const {
      dataset = 'default'
    } = options;
    return (this._labels[dataset][k] !== undefined) ? this._labels[dataset][k]: k;
  }

  downloadCSV(filename, dimensionName='_all', options={}) {
    const {
      dataset = 'default',
      labels = this._labels[dataset] || {},
    } = options;

    if (dimensionName === '_all' && !this._dimensions[dataset][dimensionName]) {
      let idx = 0;
      this.registerDimension('_all', (d) => idx++, {dataset})
    }
    else if (!this._dimensions[dataset][dimensionName]) {
      console.log('dimension not registered')
      return;
    }

    downloadCSV(
      this._dimensions[dataset][dimensionName].top(Infinity),
      filename,
      labels
    )
  }

}




export default new DashboardStore()
