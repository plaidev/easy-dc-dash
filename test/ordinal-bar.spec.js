import Vue from 'vue'
import Store from '../libs/store'
import Chart from '../libs/chart'
import {init, run} from '../libs/autorun'
import { mount, createLocalVue } from 'vue-test-utils'


import Base from '@/_coordinateGridBase.js'
import OrdinalBar from '@/ordinal-bar.vue'

const localVue = createLocalVue()
localVue.use(Base)

beforeAll(() => {
  init()
  Chart.install(Vue)

  Store.registerData([
    {d1: 'a', d2: true, v: 1},
    {d1: 'a', d2: false, v: 2},
    {d1: 'b', d2: false, v: 3},
    {d1: 'b', d2: true, v: 5},
  ])
  run()
})

describe('ordinal-bar', () => {
  const wrapper = mount(OrdinalBar, {
    localVue,
    attachToDocument: true,
    propsData: {
      dimension: 'd1',
      reduce: 'v'
    }
  })

  // console.log(wrapper.vm)
  const div = wrapper.findAll('div')
  console.log(div);
  const svg = wrapper.findAll('svg')
  console.log(svg);
  const g = wrapper.findAll('g')
  console.log(g);

  it('renders the correct bar', () => {
    expect(wrapper.hasProp('chartType', 'barChart')).toBe(true)
    expect(wrapper.hasProp('dimension', 'd1')).toBe(true)
    expect(wrapper.hasProp('reduce', 'v')).toBe(true)
  })
})