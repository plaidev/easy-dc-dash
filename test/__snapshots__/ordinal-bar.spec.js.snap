// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ordinal-bar renders the correct bar 1`] = `
<div class="outer-container" style="width: 377px; height: 233px;">
    <div class="backdrop"></div>
    <div class="card-container">
        <div class="inner-container" style="width: 377px; height: 233px;">
            <div class="container-header">
                <div class="icon-box">
                    <i class="fa fa-window-maximize"></i>
                </div>
                <!---->
            </div>
            <div class="render-area">
                <div id="test-ordinal-bar" class="krt-dc-component dc-chart" style="display: flex; position: relative; width: 100%; height: 100%;">
                    <!---->
                    <div class="reset-button">
                        <a class="reset" style="display: none;">
                            <i class="fa fa-ban">&nbsp;</i> <span>フィルター解除</span></a>
                    </div>
                    <!---->
                    <svg width="200" height="200">
                        <g>
                            <g class="grid-line horizontal" transform="translate(40,20)"></g>
                            <g class="chart-body" transform="translate(40, 20)" clip-path="url(about:blank#test-ordinal-bar-clip)">
                                <g class="stack _0">
                                    <rect class="bar" fill="#1f77b4" y="150" height="0"></rect>
                                    <rect class="bar" fill="#1f77b4" y="150" height="0"></rect>
                                    <text class="barLabel" text-anchor="middle" cursor="pointer"></text>
                                    <text class="barLabel" text-anchor="middle" cursor="pointer"></text>
                                </g>
                            </g>
                            <g class="axis x" transform="translate(40,170)">
                                <g class="tick" style="opacity: 0.000001;" transform="translate(39.310344827586206,0)">
                                    <line y2="6"></line>
                                    <text dy=".71em" style="text-anchor: middle;" y="9">a</text>
                                </g>
                                <g class="tick" style="opacity: 0.000001;" transform="translate(80.6896551724138,0)">
                                    <line y2="6"></line>
                                    <text dy=".71em" style="text-anchor: middle;" y="9">b</text>
                                </g>
                                <path class="domain"></path>
                            </g>
                            <g class="axis y" transform="translate(40,20)">
                                <path class="domain"></path>
                            </g>
                        </g>
                        <defs>
                            <clippath id="test-ordinal-bar-clip">
                                <rect width="120" height="150" transform="translate(-0, -0)"></rect>
                            </clippath>
                        </defs>
                    </svg>
                </div>
            </div>
        </div>
    </div>
</div>
`;
